domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt

```diff
@@ -1408,11 +1781,304 @@ class ExplorationProgressControllerTest {
+    )
+    playThroughPrototypeState1AndMoveToNextState()
+    playThroughPrototypeState2AndMoveToNextState()
+    moveToPreviousState()
+
+    verifyCheckpointHasCorrectStateIndex(
+      profileId,
+      TEST_EXPLORATION_ID_2,
+      stateIndex = 2
+    )
+  }
+
+  @Test
+  fun testCheckpointing_advToThirdState_prevStates_nextState_checkCheckpointHasCorrectStateIndex() {
+    subscribeToCurrentStateToAllowExplorationToLoad()
+    playExploration(
+      profileId.internalId,
+      TEST_TOPIC_ID_0,
+      TEST_STORY_ID_0,
+      TEST_EXPLORATION_ID_2,
+      isCheckpointingEnabled = true
+    )
+    playThroughPrototypeState1AndMoveToNextState()
+    playThroughPrototypeState2AndMoveToNextState()
+    moveToPreviousState()
+    moveToPreviousState()
+    moveToNextState()
+
+    verifyCheckpointHasCorrectStateIndex(
+      profileId,
+      TEST_EXPLORATION_ID_2,
+      stateIndex = 2
+    )
+  }
+
+  @Test
+  fun testCheckpointing_revealHint_checkHintIsSavedInCheckpoint() {
+    subscribeToCurrentStateToAllowExplorationToLoad()
+    playExploration(
+      profileId.internalId,
+      TEST_TOPIC_ID_0,
+      TEST_STORY_ID_0,
+      TEST_EXPLORATION_ID_2,
+      isCheckpointingEnabled = true
+    )
+    navigateToPrototypeFractionInputState()
+    submitWrongAnswerForPrototypeState2()
+
+    verify(
+      mockCurrentStateLiveDataObserver,
+      atLeastOnce()
+    ).onChanged(currentStateResultCaptor.capture())
+    assertThat(currentStateResultCaptor.value.isSuccess()).isTrue()
+    val currentState = currentStateResultCaptor.value.getOrThrow()
+
+    val result = explorationProgressController.submitHintIsRevealed(
+      state = currentState.state,
+      hintIsRevealed = true,
+      hintIndex = 0,
+    )
+    result.observeForever(mockAsyncHintObserver)
+    testCoroutineDispatchers.runCurrent()
+
+    verifyCheckpointHasCorrectHintIndex(
+      profileId,
+      TEST_EXPLORATION_ID_2,
+      indexOfRevealedHint = 0
+    )
+  }
+
+  @Test
+  fun testCheckpointing_revealSolution_checkCheckpointIsSaved() {
+    subscribeToCurrentStateToAllowExplorationToLoad()
+    playExploration(
+      profileId.internalId,
+      TEST_TOPIC_ID_0,
+      TEST_STORY_ID_0,
+      TEST_EXPLORATION_ID_2,
+      isCheckpointingEnabled = true
+    )
+    navigateToPrototypeFractionInputState()
+    submitWrongAnswerForPrototypeState2()
+
+    verify(
+      mockCurrentStateLiveDataObserver,
+      atLeastOnce()
+    ).onChanged(currentStateResultCaptor.capture())
+    assertThat(currentStateResultCaptor.value.isSuccess()).isTrue()
+    val currentState = currentStateResultCaptor.value.getOrThrow()
+
+    val result = explorationProgressController.submitSolutionIsRevealed(currentState.state)
+    result.observeForever(mockAsyncSolutionObserver)
+    testCoroutineDispatchers.runCurrent()
+
+    verifyCheckpointHasCorrectValueOfIsSolutionRevealed(
+      profileId,
+      TEST_EXPLORATION_ID_2,
+      isSolutionRevealed = true
+    )
```

Comment:
Do we need a test for the case where the player reaches 'Continue' ? I think you have some special casing for that scenario right?

Reply:
I don't think we need any additional test case for that because in the case of `Continue` we have just optimized a bit by saving one checkpoint instead of two.

Reply:
I think we should then verify that the optimization works as intended right?

Reply:
added a test that verifies that the optimization works as intended.